### ISSUE TITLE
`nextpow(2, x)` is incorrect for large numbers

### ISSUE THREAD

**Issue by yurivish** [JuliaLang/julia#45880]
The return value of `nextpow(a, x)` is [documented](https://docs.julialang.org/en/v1/base/math/#Base.nextpow) to be no less than `x`. However, `nextpow` returns negative numbers for some positive inputs.

For example:

```jl
julia> nextpow(2, Int8(100))
-128
```

This is due to an incorrect “fast path” optimization for powers of 2:

https://github.com/JuliaLang/julia/blob/fec6951a8b2e11a2560d13c1b995874bff561aa5/base/intfuncs.jl#L456-L459

This optimization causes the result to be incorrect for ~25% of all signed integers (`Int8`, `Int16`, `Int32`, `Int64`, `Int128`) and ~50% of unsigned integers (`UInt8`, `UInt16`, `UInt32` `UInt64`, `UInt128`).

```jl
julia> nextpow(2, Int16(20_000))
-32768

julia> nextpow(2, UInt16(40_000))
0x0000
```

In addition, `nextprod` hangs forever on the following invocation due to [miscomputing an internal variable](https://github.com/JuliaLang/julia/blob/master/base/combinatorics.jl#L334):

```jl
nextprod((2, 3), Int8(105))
```

Tested on Julia 1.7.3.

```
julia> versioninfo()
Julia Version 1.7.3
Commit 742b9abb4d (2022-05-06 12:58 UTC)
Platform Info:
  OS: macOS (x86_64-apple-darwin21.4.0)
  CPU: Intel(R) Core(TM) i9-9980HK CPU @ 2.40GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-12.0.1 (ORCJIT, skylake)
```

**Comment by timholy** [comment]
Isn't this just overflow?
```julia
julia> using CheckedArithmetic

julia> @check nextpow(2, Int8(100))
ERROR: InexactError: trunc(Int8, 128)
Stacktrace:
 [1] throw_inexacterror(f::Symbol, #unused#::Type{Int8}, val::Int128)
   @ Core ./boot.jl:614
 [2] checked_trunc_sint
   @ ./boot.jl:636 [inlined]
 [3] toInt8
   @ ./boot.jl:652 [inlined]
 [4] Int8
   @ ./boot.jl:761 [inlined]
 [5] convert
   @ ./number.jl:7 [inlined]
 [6] round
   @ ./int.jl:649 [inlined]
 [7] safeconvert(#unused#::Type{Int8}, x::Int128)
   @ CheckedArithmetic ~/.julia/packages/CheckedArithmetic/ThC4h/src/CheckedArithmetic.jl:218
 [8] top-level scope
   @ ~/.julia/packages/CheckedArithmetic/ThC4h/src/CheckedArithmetic.jl:119
```

Mathematically `x + 1` is also bigger than `x` but that's not always true in the face of overflow.

**Comment by yurivish** [comment]
Yes, it is. However, for invocations that don't hit the special case, ie. powers other than 2, the result is correct: 

```jl
julia> nextpow(3, Int8(100))
243
```

**Comment by timholy** [comment]
Interesting. I wonder if that's the real problem. Arguably
```julia
julia> convert(Int8, nextpow(3, Int8(100)))
ERROR: InexactError: trunc(Int8, 243)
Stacktrace:
 [1] throw_inexacterror(f::Symbol, #unused#::Type{Int8}, val::Int64)
   @ Core ./boot.jl:614
 [2] checked_trunc_sint
   @ ./boot.jl:636 [inlined]
 [3] toInt8
   @ ./boot.jl:651 [inlined]
 [4] Int8
   @ ./boot.jl:761 [inlined]
 [5] convert(#unused#::Type{Int8}, x::Int64)
   @ Base ./number.jl:7
 [6] top-level scope
   @ REPL[6]:1
```
is the actual problem. (Certainly the inconsistency is not good.)

I guess the real question is whether we want to use modular arithmetic or not.

**Comment by yurivish** [comment]
The documentation implies that the error behavior is the correct one, since it says that the result should never be less than its input.

I think this means that relative to the documentation the current behavior on large integers is incorrect in general, not just for the special-cased power:

```jl
julia> nextpow(3, typemax(Int) / 2 + 1)
-6289078614652622815
```

**Comment by JeffBezanson** [comment]
Thank you for all these recent bug reports!

Ideally a function like `sum` just calls `+`, so that it doesn't need a separate decision on what to do on overflow.

I'm not 100% sure what `nextpow` should do, but I think it's complex and expensive enough to warrant an overflow check.

**Comment by oscardssmith** [comment]
`nextpow2` specifically shouldn't have one (IMO).  It's a function that is very useful for doing bitmath, and it's important that it is fast.

**Comment by gbaraldi** [comment]
In this case then I guess the docs should stress that for some cases integer overflow can happen.

**Comment by Moelf** [comment]
@yurivish I think your reading of the documentation might be too pedantic sometimes,
>The documentation implies ... the result should never be less than its input.

at this rate we will need to add "this function doesn't protect against overflow" for almost every single function in Julia Base; for this specific case, I can see the argument is that power of non-2 base has different behavior, but fundamentally, reading documentation as mathematical statement is just not practical:
```julia
julia> nextpow(3, typemax(Int64)÷2)
-6289078614652622815
```

**Comment by yurivish** [comment]
The purpose of documentation is to provide readers with the knowledge they need to use functions correctly, which includes their behavior in exceptional cases. 

I agree that Julia’s integer overflow behavior is documented, as is `sum`’s type promotion behavior.

**Comment by nlw0** [comment]
Regardless of the documentation, the function signature takes `::Real` numbers, and in the body of the function below there are actually overflow checks. And because of the `isa(x, Integer)`, we can even observe something like:

```
julia> nextpow(2.0, Int8(100))
-128
```

It's not that intuitive that even with one of the arguments being floating-point, we still get an overflow error. I guess it does make sense because it's all about the exponent.

Another relevant test in the code itself says that "`x`must be positive", what is also kind of ironic. I don't really know what to think about this function. Maybe if we're fine with the overflows we could at least have a separate implementation for x::Integer that skips all the testing?

**Comment by adienes** [comment]
I believe anytime that internal overflow is allowed to mangle results away from the otherwise documented behavior, it should be documented that is a possibility. the docs only say

> Arithmetic operations with Julia's integer types inherently perform [modular arithmetic](https://en.wikipedia.org/wiki/Modular_arithmetic), mirroring the characteristics of integer arithmetic on modern computer hardware. In scenarios where overflow is a possibility, it is crucial to explicitly check for wraparound effects that can result from such overflows

which only describe the behavior of arithmetic operations, aka things like `*`, `+`, `-`,`^` and I do not read that as giving julia license to be wrong in any function that happens to use arithmetic internally.

furthermore, this particular issue is not "just overflow" because it hits float inputs as well
```
julia> nextpow(2.0, typemax(Int))
-9223372036854775808
```

**Comment by vtjnash** [comment]
> which only describe the behavior of arithmetic operations, aka things like *, +, -,^ and I do not read that as giving julia license to be wrong in any function that happens to use arithmetic internally.

Modular arithmetic is a consistent ring and isn't "wrong" in any sense compared to other number systems. There's a separate question of whether this function does actually implement the operation in the right modulus base, and whether the documentation should be improved to avoid the phrasing of the result being always greater.

**Comment by adienes** [comment]
`Z/nZ` is not an ordered ring so I think that for any definition of a function relying on an order (such as `nextpow` does) the appeals to modular arithmetic fall a little flat IMO. In the case of `nextpow` on integers, I can see the argument that the overflow is intentional, but on floats `nextpow(2.0, typemax(Int))` is clearly wrong

my statement above I mostly mean to highlight that just because arithmetic is allowed to overflow doesn't necessarily mean that whatever _internal_ overflow happens is allowed to change the result to be implementation-dependent. https://github.com/JuliaLang/julia/issues/59684 is probably a more obvious example of an incorrect result due to internal overflow.

**Comment by vtjnash** [comment]
It is defined here numerically (in the code) as `a ^ ceillog_a(x)` or `a ^ (floorlog_a(x) + 1)` in the numeric ring R (for Int, UInt, Float, etc.)

Yes, that is a bug in consistency of the function and fast path. There is a PR to fix that here:
https://github.com/JuliaLang/julia/pull/49672
I've made an alternative proposal to that of returning `oftype(x)` instead of `promote(a, x)`

**Comment by vtjnash** [comment]
Trivia: nextprod is quite old: e642271963299fef907d0a56517d680021ef9b4b, and it sounds like it might be the unbounded knapsack problem given there was an ILP solver written for it (https://github.com/JuliaLang/julia/pull/981#issuecomment-6633648)

**Comment by adienes** [comment]
I don't yet understand how returning `oftype(x)` will resolve the `nextpow(2.0, typemax(Int))` case; that looks like it would still be type unstable on float inputs

> it might be the unbounded knapsack problem

if we're intentionally allowing modular arithmetic for integer inputs, it's the discrete log problem! consider
```
julia> Int8(14)^4
16
```
should `nextpow(Int8(14), 15) == 16` ?
